### DISCUSSION TITLE
SeleniumBase behaving differently on Linux then on Windows

### DISCUSSION THREAD

**Discussion by FindingBen** [seleniumbase/SeleniumBase#3161] [Q&A] [answered]
Basically I am logging into snapchat with seleniumbase and it works perfectly in headless mode on windows, the way it works is I have 3 steps. 1st step inputs a username, 2nd step it clicks the next button then it goes to password input, then 3rd step it inputs password and finally 4th step it clicks button to sign in.

For some reason on Linux it reaches the password page , it inputs the text but when I click the next button it doesnt log user in. And the page which I am suppose to be on is not present. I have bunch of print methods which are giving me info about the pages so thats how I can see whats currently present.

Here is my code below:

`
  with SB(headless=True,undetectable=True) as sb:
          print("AS")
          try:
              print('Ok reached here')
              sb.driver.uc_open_with_reconnect("https://accounts.snapchat.com/accounts/v2/login", 4)
              sb.sleep(4)
              print('Hey am here')
              sb.execute_script("""
                      var elements = document.querySelectorAll("section[class*='sdsm-modal sdsm-modal-blocking background--white css-xqeail']");
                      elements.forEach(function(element) {
                          element.parentNode.removeChild(element);
                      });
              """)
  
              sb.driver.wait_for_element_visible("input[name='accountIdentifier']", timeout=10)
  
              sb.driver.type("input[name='accountIdentifier']", username)
              dd=sb.driver.get_title()
              print("USERNAME",dd)
              sb.driver.sleep(3)
              sb.driver.wait_for_element_visible("//button[contains(text(), 'Next')]", timeout=10)
  
              sb.driver.sleep(2)  # Give some time for the scrolling to complete
              sb.driver.click("//button[contains(text(), 'Next')]")
              sb.driver.sleep(10)
              sb.driver.execute_script("""
                      var elements = document.querySelectorAll("section[class*='sdsm-modal sdsm-modal-blocking background--white css-xqeail']");
                      elements.forEach(function(element) {
                          element.parentNode.removeChild(element);
                      });
              """)
            sb.driver.wait_for_element_visible("input[name='password']", timeout=10)
            sb.driver.type("input[name='password']", password)
            
            sb.driver.sleep(2)
            sss = sb.driver.get_title()
            print("PASS:",sss)
            
            sb.driver.wait_for_element_visible("//button[contains(text(), 'Next')]", timeout=15)
            
            sb.driver.submit("//button[contains(text(), 'Next')]")
            
            print('found 4')
            sb.driver.sleep(15)
            try:
                error_message = sb.driver.wait_for_element_visible("//div[id='error_message']",10)
                if error_message:
                    print("Error message found:", error_message.text)
                else:
                    print("No error message found.")
            except Exception as e:
                print("Error message not found:", e)
            saass = sb.driver.get_title()
            print("MAY:",saass)
            sb.driver.uc_open_with_reconnect(invatation_url,5)
            sb.driver.sleep(15)
            saapagess = sb.driver.get_title()
            print('MAYBE:',saapagess)
            try:
                close = sb.driver.wait_for_element_visible("//button[span[contains(text(), 'Accept All')]]", timeout=10)
                close.click()
                sb.driver.sleep(10)
                print('CLICKED!')
            except Exception as e:
                print('Error', e)
            print('almost..')
            ss = sb.driver.get_title()
            print(ss)
            sb.driver.wait_for_element_visible("//button[span[contains(text(), 'Accept')]]", timeout=10)
            sb.driver.click("//button[span[contains(text(), 'Accept')]]")
            sb.driver.sleep(7)
            sb.driver.quit()
            print("Logged in successfully!")
        finally:
            sb.driver.quit()
`

So the web driver gets initialized, it passed the first two steps..but then it doesn't pass the 3rd one..it works well on windows so I know its not the element error or whatever, also my credentials are hardcoded for this case so that is correct as well.

I know this is not a new case but I would really like some input on this. Any help is appreciated!

**Comment by mdmintz** [comment, marked as answer]
UC Mode doesn't really support headless mode anymore because UC Mode uses `PyAutoGUI` for lots of things, which doesn't work with a headless browser. (You don't need headless mode on Linux anymore because of the special virtual display that's included.)

Instead of using headless mode on Linux, use `xvfb=True` to activate the virtual display.

If you encounter a CF CAPTCHA, you'll need to use `sb.uc_gui_click_captcha()`.

If you run your script with `--debug` on the command-line, it prints out the coordinates clicked via `sb.uc_gui_click_captcha()`. With those coordinates, you can call `sb.uc_gui_click_x_y(x, y)` from a disconnected state (discussed in the [UC Mode docs](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/uc_mode.md)), which avoids an extra page refresh that occurs with `sb.uc_gui_click_captcha()`. Since Selenium may be detected when calculating the coordinates of the CAPTCHA checkbox, a page refresh occurs before clicking the CAPTCHA checkbox to avoid detection.

**Comment by FindingBen** [comment]
So I tried the xvfb before 

and I am getting error of X11 display failed! Will use regular xvfb and then it just freezes..

![Detailed error](https://github.com/user-attachments/assets/a56eff18-89bf-4220-bd61-195268fcf153)

and my code where I use it:

` with SB(xvfb=True,undetectable=True) as sb:
        print("AS")
        try:
            print('Ok reached here')
            sb.driver.uc_open_with_reconnect("https://accounts.snapchat.com/accounts/v2/login", 4)
            sb.sleep(4)
            print('Hey am here')
            sb.execute_script("""
                    var elements = document.querySelectorAll("section[class*='sdsm-modal sdsm-modal-blocking background--white css-xqeail']");
                    elements.forEach(function(element) {
                        element.parentNode.removeChild(element);
                    });
            """)`


But either way, you say UC mode doesnt support headless but my driver still gets initiated and it goes to type in the username and goes to password page..so I would assume that driver wouldn't even be initiated if that was the case or am I wrong?

**Reply by mdmintz** [reply]
You need a headed browser for the `PyAutoGUI` actions. Headless mode is fine for regular browsing. However, a headless browser makes you more likely to get detected by anti-bot systems, where then you'd need the headed browser for the `PyAutoGUI` actions.

If you get `X11 display failed!`, it means your version of Linux is missing `X11` dependencies (in which case you would need to install those separately, and probably turn to Stack Overflow, etc, for how to do that). Most flavors of Linux, (such as GitHub Actions Linux),  already have all the necessary `X11` dependencies installed.
